kcidb/monitor/test_output.py

```diff
@@ -85,3 +85,57 @@ def test_min(self):
         self.assertEqual('utf-8', html.get_content_charset())
         content = html.get_payload()
         self.assertIn("We detected a new revision!", content)
+
+    def test_subject_and_body_length(self):
+        """Check subject and body length of Notification """
+
+        db_client = db.Client("sqlite::memory:")
+        db_client.init()
+        oo_client = oo.Client(db_client)
+        db_client.load({
+            "version": self.version,
+            "checkouts": [
+                {
+                    "contacts": [
+                        "rdma-dev-team@redhat.com"
+                    ],
+                    "start_time": "2020-03-02T15:16:15.790000+00:00",
+                    "git_repository_branch": "wip/jgg-for-next",
+                    "git_commit_hash": "5e29d1443c46b6ca70a4c940a67e8c09f05dcb7e",
+                    "patchset_hash": "",
+                    "git_repository_url": "https://git.kernel.org/pub/scm/linux/kernel/git/arm64/linux.git",
+                    "misc": {
+                        "pipeline_id": 467715
+                    },
+                    "id": "origin:1",
+                    "origin": "origin",
+                    "patchset_files": [],
+                    "valid": True,
+                },
+            ],
+            "builds": [
+                                {
+                    "id": "kernelci:kernelci.org:619fecfc6d932a49b5f2efb0",
+                    "checkout_id": "origin:1",
+                    "origin": "origin",
+                    "comment": "v5.16-rc2-19-g383a44aec91c",
+                    "start_time": "2021-11-25T20:07:24.499000+00:00",
+                    "duration": 486.695294857,
+                    "architecture": "x86_64",
+                    "config_name": "x86_64_defconfig+x86-chromebook+kselftest",
+                    "valid": True
+                },
+            ]
+        })
+        oo_data = oo_client.query(orm.Pattern.parse(">*#"))
+        notification = monitor.match(oo_data)[0]
+        notification.obj.git_commit_name = "Set an invalid data with length greater " \
+                                           "than notification subject max length" * 40
+        notification.obj.git_commit_hash = "Set an invalid commit hash with length greater " \
+                                           "than notification body max length" * 250
+        self.assertEqual(notification.id,
+                         'mark_brown:revision:'
+                         'KCc1ZTI5ZDE0NDNjNDZiNmNhNzBhNGM5NDBhNj'
+                         'dlOGMwOWYwNWRjYjdlJywgJycp:')
+        message = notification.render()
+        self.assertIsInstance(message, email.message.EmailMessage)
```

Comment:
This is a pretty good start. However, you don't actually check that either the subject or the body length was (not)limited.

To test this well, you would need to check these for both subject and body:

- [ ] Zero-length NotificationMessage subject/body produces zero length subject and body in the email, without failing
- [ ] Under-limit length subject/body are left intact in the email
- [ ] Exact-limit length subject/body are left intact in the email, with all characters being single-byte in UTF-8
- [ ] Exact-limit length subject/body are left intact in the email, with at least one character being multi-byte in UTF-8 (verifying that we're limiting by characters, not by bytes)
- [ ] Over-limit length subject/body are trimmed to the exact limit with the last character replaced with the scissors emoji in the email

Also, you don't really need to involve subscriptions to create a notification. Having them involved makes it harder to craft the notifications you need for testing. Instead you can create notification objects directly. Something like this:
```python
        db_client = db.Client("sqlite::memory:")
        db_client.init()
        db_client.load(io_data)
        oo_client = oo.Client(db_client)
        revision = oo_client.query(orm.Pattern.parse(">revision#"))["revision"][0]
        message = NotificationMessage(
            ["kernelci-results-staging@groups.io"],
            subject="JINJA2_TEMPLATE_FOR_YOUR_SUBJECT",
            body="JINJA2_TEMPLATE_FOR_YOUR_BODY"
        )
        notification = Notification(revision, "test_subscription", message)
        email = notification.email()
```